server/datastore/mysql/queries.go

```diff
@@ -136,9 +247,33 @@ func (ds *Datastore) SaveQuery(ctx context.Context, q *fleet.Query) error {
 	return nil
 }
 
-// DeleteQuery deletes Query identified by Query.ID.
-func (ds *Datastore) DeleteQuery(ctx context.Context, name string) error {
-	return ds.deleteEntityByName(ctx, queriesTable, name)
+func (ds *Datastore) DeleteQuery(
+	ctx context.Context,
+	teamID *uint,
+	name string,
+) error {
+	stmt := "DELETE FROM queries WHERE name = ?"
+
+	args := []interface{}{name}
+	whereClause := " AND team_id_char = ''"
+	if teamID != nil {
+		args = append(args, fmt.Sprint(*teamID))
+		whereClause = " AND team_id_char = ?"
+	}
+	stmt += whereClause
+
+	result, err := ds.writer(ctx).ExecContext(ctx, stmt, args...)
+	if err != nil {
+		if isMySQLForeignKey(err) {
```

Comment:
Any foreign keys that will fail with the new changes?

Reply:
This was only used here - https://github.com/fleetdm/fleet/pull/12673/files#diff-5340900423eb6bc2b31a2e535752ae68fe8ce78f009b5aae7988af7fb38fc4b1R81. For old packs, the query foreign key constraint was based on the query name. It is assumed that if a pack references a query, is because the query lives in the 'global' space (`team_id = NULL`).

Reply:
Gotcha.

> It is assumed that if a pack references a query, is because the query lives in the 'global' space (team_id = NULL).

Do you mean `scheduled_queries`? (An entry in `scheduled_queries` references a foreign entry in `queries`).

So, the user cannot delete global queries that are used by 2017 packs, right? The user must first delete the 2017 pack?

Reply:
> Do you mean `scheduled_queries`? (An entry in `scheduled_queries` references a foreign entry in `queries`).
> 

Yup, sorry - the relationship chain goes like `packs` -> `scheduled_queries` -> `queries`

> So, the user cannot delete global queries that are used by 2017 packs, right? The user must first delete the 2017 pack?

Ah, yeah good catch, I see what you mean ... I think we will need to add a `query_id ` FK to `scheduled_queries` to preserve consistency then. All `query_id` values in `scheduled_queries` should be correct, right? Also, just to double check, after this is deployed, users won't be able to 'ingest' 2017 packs directly, right?

Reply:
Maybe we can add a new `query` column (for the SQL) to `scheduled_queries` as part of the migration? To split `queries` and `scheduled_queries` completely? (If it helps us.)

Reply:
> Also, just to double check, after this is deployed, users won't be able to 'ingest' 2017 packs directly, right?

My guess is that they still can? (to not break compatibility) Let's double check with product. @zhumo @rachaelshaw

Reply:
I just checked and currently the user can delete queries even if they are referenced by scheduled queries or 2017 packs. So that should still be the case after we release this.

Reply:
Thanks Lucas - updated the FK constraint in https://github.com/fleetdm/fleet/pull/12673/commits/19d3c29fe0091cbc081de6e091451243ae0cd117 to allow for that

Reply:
Yes! Until we give people the tools to migrate from 2017 packs and convert community packs into fleet, we should allow you to continue as you were before.